src/validator.cpp

```diff
@@ -1635,6 +1708,530 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
+        addMathMLIssue("Math has a '" + node->name() + "' element without a MathML child.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasAtLeastOneMathmlChild(const XmlNodePtr &node,
+                                                        const ComponentPtr &component)
+{
+    if (mathmlChildCount(node) < 1) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without at least one MathML child.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasTwoMathmlChildren(const XmlNodePtr &node,
+                                                    const ComponentPtr &component)
+{
+    if (mathmlChildCount(node) != 2) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without two MathML children.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+bool Validator::ValidatorImpl::hasOneOrTwoMathmlChildren(const XmlNodePtr &node,
+                                                         const ComponentPtr &component)
+{
+    auto childCount = mathmlChildCount(node);
+
+    if ((childCount != 1) && (childCount != 2)) {
+        addMathMLIssue("Math has a '" + node->name() + "' element without one or two MathML children.",
+                       Issue::ReferenceRule::MATH_MATHML,
+                       component);
+
+        return false;
+    }
+
+    return true;
+}
+
+void Validator::ValidatorImpl::validateMathMLElementsChildrenAndSiblings(const XmlNodePtr &node,
+                                                                         const ComponentPtr &component)
+{
+    // Check the current node against the MathML elements listed in
+    // supportedMathMLElements.
+    // Note: we don't have to do anything for 'true', 'false, 'exponentiale',
+    //       'pi', 'infinity', and 'notanumber' since they must be empty and if
+    //       they are not then this will be caught by the DTD.
+
+    // Basic content elements.
+
+    if (node->isMathmlElement("apply")) {
+        if (hasAtLeastOneMathmlChild(node, component)) {
+            for (size_t i = 0, iMax = mathmlChildCount(node); i < iMax; ++i) {
+                validateMathMLElementsChildrenAndSiblings(mathmlChildNode(node, i), component);
+            }
+        }
+
+        // Assignment, and relational and logical operators.
+
+    } else if (node->isMathmlElement("eq")
+               || node->isMathmlElement("neq")
+               || node->isMathmlElement("lt")
+               || node->isMathmlElement("leq")
+               || node->isMathmlElement("gt")
+               || node->isMathmlElement("geq")) {
+        auto parentNode = node->parent();
```

Comment:
eq, leq, lt, geq, gt are n-ary operators (followed by zero or more child elements) and neq is a binary operator (followed by exactly two child elements). I don't recall the CellML spec putting additional restrictions on the n-ary operators...but I could easily be wrong.

Reply:
> eq, leq, lt, geq, gt are n-ary operators (followed by zero or more child elements) and neq is a binary operator (followed by exactly two child elements). I don't recall the CellML spec putting additional restrictions on the n-ary operators...but I could easily be wrong.

Argh, it is clearly a leftover from my COR code! I have just double checked the CellML specs and, as I expected, we don't have any restrictions on those operators.

> I'm also starting to think that maybe this validation is designed to validate against what the code generation can work with, rather than strictly validating against the MathML spec...

Apart from my oversight regarding the `minus` and `divide` operators, I used the MathML specifications to validate things.

Now, I clearly overlooked the type of operators `eq`, `neq`, `lt`, `leq`, `gt`, and `geq` are. `neq` is indeed a binary operator while the others are n-ary operators as confirmed [here](https://www.w3.org/TR/MathML2/chapter4.html#contm.relation).

I am, however, confused as how we can interprete an n-ary operator. If we consider `eq`, it means that we could have:
- No operand, i.e.
```xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
    <apply>
        <eq/>
    </apply>
</math>
```
What do/can we do with that?
- One operand, e.g.
```xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
    <apply>
        <eq/>
        <ci>a</ci>
    </apply>
</math>
```
What do/can we do with that?
- Two operands, e.g.
```xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
    <apply>
        <eq/>
        <ci>a</ci>
        <ci>b</ci>
    </apply>
</math>
```
That's what we support and we interpret it as meaning `a = b` if `eq` is the root element of a MathML equation or `a == b` otherwise.
- Three or more operands, e.g.
```xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
    <apply>
        <eq/>
        <ci>a</ci>
        <ci>b</ci>
        <ci>c</ci>
        <ci>d</ci>
        <ci>e</ci>
    </apply>
</math>
```
I understand this as meaning `a = b = c = d = e` or, once code has been generated for it, as `a = b`, `b = c`, `c = d`, and `d = e`. That is, if `eq` is the root element of a MathML equation. If iis not then I understand this as meaning `a == b == c == d == e` or `(a == b) && (b == c) && (c == d) && (d == e)`.

I guess that in COR, I went for the easiest solution and considered those operators as binary operators, and did the same here.

@nickerso and @hsorby... views?

Reply:
yeah, this was why I was wondering if we really should be validating against what libCellML can handle rather than the MathML spec directly. I'm happy to leave this "as is" and punt it to a future issue to handle the validation and code generation for these cases when a need is demonstrated. We can be fairly confident in knowing that these are currently only used in the binary manner for now....I think :)

Reply:
Yeah, I think that after more than 20 years, it's pretty safe to say that people have been using those operators in a binary manner. :)

Reply:
I'll resolve this thread for now and have added #1139 so we might remember the issue in future. @hsorby - feel free to reopen if you think we need to address this in this PR.